src/System.IO.FileSystem/src/System/IO/DosMatcher.cs

```diff
@@ -0,0 +1,307 @@
+            int currentState;
+            bool nameFinished = false;
+
+            while (!nameFinished)
+            {
+                if (nameOffset < name.Length)
+                {
+                    // Not at the end of the name. Grab the current character and move the offset forward.
+                    nameChar = name[nameOffset++];
+                }
+                else
+                {
+                    // At the end of the name. If the expression is exhausted, exit.
+                    if (priorMatches[matchCount - 1] == maxState)
+                        break;
+
+                    nameFinished = true;
+                }
+
+                // Now, for each of the previous stored expression matches, see what
+                // we can do with this name character.
+                priorMatch = 0;
+                currentMatch = 0;
+                priorMatchCount = 0;
+
+                while (priorMatch < matchCount)
+                {
+                    // We have to carry on our expression analysis as far as possible for each
+                    // character of name, so we loop here until the expression stops matching.
+
+                    expressionOffset = (priorMatches[priorMatch++] + 1) / 2;
+
+                    while (expressionOffset < expression.Length)
+                    {
+                        currentState = expressionOffset * 2;
+                        expressionChar = expression[expressionOffset];
+
+                        // We may be about to exhaust the local space for matches,
+                        // so we have to reallocate if this is the case.
+                        if (currentMatch >= currentMatches.Length - 2)
+                        {
+                            int newSize = currentMatches.Length * 2;
+                            temp = new Span<int>(new int[newSize]);
+                            currentMatches.CopyTo(temp);
+                            currentMatches = temp;
+
+                            temp = new Span<int>(new int[newSize]);
+                            priorMatches.CopyTo(temp);
+                            priorMatches = temp;
+                        }
+
+                        if (expressionChar == '*')
+                        {
+                            // '*' matches any character zero or more times.
+                            goto MatchZeroOrMore;
+                        }
+                        else if (expressionChar == '<')
+                        {
+                            // '<' (DOS_STAR) matches any character except '.' zero or more times.
+
+                            // If we are at a period, determine if we are allowed to
+                            // consume it, i.e. make sure it is not the last one.
+
+                            bool notLastPeriod = false;
+                            if (!nameFinished && nameChar == '.')
+                                for (int offset = nameOffset; offset < name.Length; offset++)
```

Comment:
Nit: please use braces for both the if and for bodies

Reply:
Ok